lib/prism/parse_result.rb

```diff
@@ -21,44 +21,87 @@ def initialize(source, start_line = 1, offsets = compute_offsets(source))
       @source = source
       @start_line = start_line
       @offsets = offsets
+      @character_offsets = nil
     end
 
     # Perform a byteslice on the source code using the given byte offset and
     # byte length.
-    def slice(offset, length)
-      source.byteslice(offset, length)
+    def slice(byte_offset, length)
+      source.byteslice(byte_offset, length)
     end
 
     # Binary search through the offsets to find the line number for the given
     # byte offset.
-    def line(value)
-      start_line + find_line(value)
+    def line(byte_offset)
+      start_line + find_line(byte_offset)
     end
 
     # Return the byte offset of the start of the line corresponding to the given
     # byte offset.
-    def line_offset(value)
-      offsets[find_line(value)]
+    def line_start(byte_offset)
+      offsets[find_line(byte_offset)]
     end
 
     # Return the column number for the given byte offset.
-    def column(value)
-      value - offsets[find_line(value)]
+    def column(byte_offset)
+      byte_offset - line_start(byte_offset)
+    end
+
+    # Return the character offset for the given byte offset.
+    def character(byte_offset)
+      character_offsets[byte_offset] ||= find_character_offset(byte_offset)
+    end
+
+    # Return the column number in characters for the given byte offset.
+    def character_column(byte_offset)
+      character(byte_offset) - character(line_start(byte_offset))
     end
 
     private
 
+    # A lazily-computed cache of byte offset => character offset mappings.
+    def character_offsets
+      @character_offsets ||= { 0 => 0 }
+    end
```

Comment:
I'm not sure this would pay off because it could become pretty large for a big file and consume N bytes for every byte in the source.
N seems to be about 60 for integer entries in a Hash:
```
ruby --disable-gems -e 'h={}; 1.times {|i| h[i]=i}'
12MB
ruby --disable-gems -e 'h={}; 10_000_000.times {|i| h[i]=i}'
588MB
irb(main):003:0> 576 * 1024**2 / 10_000_000.0
=> 60.3979776
```

I think it's better to leave this uncached for now, or do one cache entry per line.

Reply:
Fair point!